### ISSUE TITLE
Redo buffer gets deleted upon Save As

### ISSUE THREAD

**Issue by aardappel** [Microsoft/vscode#17977]
- VSCode Version: 1.8.1
- OS Version: Windows 10

Steps to Reproduce:

1. Undo last edits.
2. Save As.
3. Try Redo to re-instate last edits, but nothing happens.

This is a workflow I have done in many other editors where I realize I want to keep an old version of something I am working on, just in case I screw up. To my dismay in VSCode it made me lose all my edits.

**Comment by bpasero** [comment]
@alexandrudima how would I be able to move an existing undo/redo buffer to a new model? I already have the view state that I can preserve and put to the new resource, it seems to me the undo/redo buffer is similar.

Probably related to https://github.com/Microsoft/vscode/issues/7536

**Comment by alexandrudima** [comment]
@bpasero Aha, I didn't realize we create a new model when using Save As. Presumably we create a new model because the `uri` of the model has changed?

Perhaps we could have a way to `rename` a model.

**Comment by bpasero** [comment]
@alexandrudima yes, it ends up being an entire new model. I actually have a very similar problem when you save an untitled file, because then the scheme changes from `untitled://` to `file://` and I have to carry over all UI state to the new model. 

Maybe instead of rename we could have a clone method where I can give a new URI and all state is cloned? It could be that someone is holding a reference to the original model and wants to keep that. In the end "Save As..." is not actually deleting the original file on disk, just creating a new one.

**Comment by alexandrudima** [comment]
:+1:  Cloning is better than renaming, no piece of code is prepared for a model to get a different uri all of a sudden.

But not all model state is ok to clone:
* undo/redo buffer is probably safe to clone (the only weirdness being that the model needs to begin with the same version id as the cloned model and not with a version id of 1, since the undo/redo manager holds on to some version ids.)
* tokenization state is probably also safe to clone, but not that important.
* probably not a good idea to clone decorations/markers.

I think in the end, only the undo/redo state needs to be cloned. We have another issue tracking the same thing, but that one would include serialization as well.

**Comment by bpasero** [comment]
@alexandrudima yeah so I am fine closing this in favour of the other if you want.

**Comment by MagnusOxlund** [comment]
An alternative solution would be to just leave the original file open along with its undo history, or don't even open the newly saved file. Other editors do not open the new file when "saved as", and I think that may be one way people develop this workflow.

**Comment by MagnusOxlund** [comment]
Don't the implications of lost work warrant a higher priority for this issue? I take it that's why #7169 received the `important` label back in 2016, although it's also now sitting on the backlog.

**Comment by gwagener** [comment]
I noticed this today when I renamed a file in the explorer and the undo buffer was lost. Just in case, I tried renaming it back to what it had been and was able to undo again. Based on the previous discussion this makes sense as it will be using the model for the original URI again. I doubt this adds any insights for the developers, but it might be useful for a user who comes across this issue.

**Comment by bpasero** [comment]
@alexdima given the work that went into `IUndoRedoService` in the past months, especially the support for triggering undo/redo even after a model was disposed and reopened again, I wonder if this makes it any easier to preserve the undo/redo stack when saving the untitled file to disk.

Specifically these methods seems interesting: 

https://github.com/microsoft/vscode/blob/ee1655a82ebdfd38bf8792088a6602c69f7bbd94/src/vs/platform/undoRedo/common/undoRedo.ts#L156-L163

Is it possible for me to capture the undo/redo stack right before saving the untitled file and then somehow restore the stack on the target resource? This would properly require the `restoreSnapshot` method to optionally accept the `URI` it is responsible for.

**Comment by alexdima** [comment]
:+1: We are definitely not too far from supporting this, but there is still some work that needs to be done to correctly "transplant" `SingleModelEditStackElement` and `MultiModelEditStackElement` to a different `URI` and then seed that in the `IUndoRedoService`.

The methods `createSnapshot`/`restoreSnapshot` are not suitable for this, `createSnapshot` simply records ids associated with `IUndoRedoStackElements` and then `restoreSnapshot` discards extra stack elements.
